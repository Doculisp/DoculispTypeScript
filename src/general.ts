import { IRegisterable } from "./types.containers";
import { IFail, ILocation, ISuccess, IUtil } from "./types.general";

class Location implements ILocation {
    private readonly _line: number;
    private readonly _char: number;
    
    constructor(line: number, char: number) {
        this._line = line;
        this._char = char;
    }

    public get line(): number {
        return this._line;
    }

    public get char(): number {
        return this._char;
    }

    toString(): string {
        return `{ line: ${this._line}, char: ${this._char} }`;
    }

    asJson() {
        return { line: this._line, char: this._char };
    }
}

function buildGeneral(): IUtil {
    function ok<T>(successfulValue: T) : ISuccess<T> {
        return {
            value: successfulValue,
            success: true,
        };
    };

    function fail(message: string, documentPath: string) : IFail {
        return {
            message,
            documentPath,
            success: false,
        };
    };

    function location(line: number, char: number): ILocation {
        return new Location(line, char);
    }

    return {
        ok,
        fail,
        location,
    };
}

const astParser: IRegisterable = {
    builder: () => buildGeneral(),
    name: 'util',
    singleton: true,
    dependencies: []
};

export {
    astParser,
};