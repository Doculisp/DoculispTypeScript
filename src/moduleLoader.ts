import { IContainer } from "./types/types.containers";
import path from "node:path";
import { globSync } from 'glob';
import { container } from "./container";

async function importModules(container: IContainer, moduleNames: string[]) {
    moduleNames.forEach(n => {
        if(n.endsWith('index')) {
            return;
        }

        import(n).then(module => {
            const keys = Object.keys(module);
            keys
                .filter(key => {
                    let value = module[key];
                    return !!value?.name && !!value?.builder;
                })
                .forEach(key => container.register(module[key]));
        });
    });
}

async function loadModules(container: IContainer) {
    let modulePath = __dirname;
    if(modulePath.endsWith('src')) {
        modulePath = path.resolve(path.join(modulePath, '../dist'));
    }

    if(!modulePath.endsWith('dist')) {
        modulePath = path.resolve(path.join(modulePath, './dist'));
    }

    const globPattern = modulePath + '/**/*.js';
    let moduleNames = (
        globSync(globPattern)
        .map(p => {
            let r = path.resolve(p);

            return r.replace('.js', '').replaceAll('\\', '/');
        })
    );

    return importModules(container, moduleNames).then(() => container);
}

const containerPromise = loadModules(container);

export {
    containerPromise
}