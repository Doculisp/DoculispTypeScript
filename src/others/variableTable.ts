import { IDictionary, IRegisterable } from "../types/types.containers";
import { IVariableExists, IVariableString, IVariableTable, IVariableTestable, Savable } from "../types/types.variableTable";

function buildVariableTable(): IVariableTestable {
    class VariableTable implements IVariableTable, IVariableTestable {
        private table: IDictionary<Savable> = {};
        private parent: IVariableTable | false = false;

        constructor();
        constructor(baseTable: (IVariableTable & IVariableTestable));
        constructor(baseTable?: (IVariableTable & IVariableTestable)) {
            if(!baseTable) {
                return;
            }

            const baseKeys = baseTable.getKeys();

            baseKeys.forEach(key => this.table[key] = baseTable.getValue(key) as Savable);
            this.parent = baseTable;
        }

        createChild(): IVariableTable {
            return new VariableTable(this);
        }
        
        addValue<T extends Savable>(key: string, value: T): IVariableExists {
            this.table[key] = value;
            return this;
        }

        addValueToStringList(key: string, value: IVariableString): IVariableExists {
            if(this.table[key]) {
                let tableValue = this.table[key] as Savable;
                if(tableValue.type === 'variable-array-string')
                    if(!tableValue.value.some(t => t.value == value.value)) {
                        tableValue.value.push(value);
                    }
            }
            else {
                this.table[key] = { value: [value], type: 'variable-array-string' };
            }
            return this;
        }

        private parentHasKey(key: string): boolean {
            return !!this.parent && !!this.parent.hasKey(key);
        }

        hasKey(key: string): boolean {
            return !!this.table[key] || this.parentHasKey(key);
        }

        private getValueFromParent<T extends Savable>(key: string): T | false {
            return !!this.parent ? this.parent.getValue(key) : false;
        }

        getValue<T extends Savable>(key: string): T | false {
            if(!!this.table[key]) {
                return this.table[key] as T;
            }

            return this.getValueFromParent(key);
        }

        asJson(): any {
            const keys = Object.keys(this.table);
            const ret: IDictionary<any> = {};

            keys.forEach(key => ret[key] = this.table[key]);

            return ret;
        }

        clear(): IVariableTestable {
            const keys = Object.keys(this.table);

            keys.forEach(key => delete this.table[key]);

            return this;
        }

        private addValueToParent<T extends Savable>(key: string, value: T): void {
            if(this.parent) {
                this.parent.addGlobalValue(key, value);
            }
        }

        addGlobalValue<T extends Savable>(key: string, value: T): IVariableExists {
            this.addValueToParent(key, value);
            this.addValue(key, value);
            return this;
        }

        addGlobalValueToStringList(key: string, value: IVariableString): IVariableExists {
            if(this.parent) {
                this.parent.addGlobalValueToStringList(key, value);
            }

            this.addValueToStringList(key, value);

            return this;
        }

        private getParentKeys(): string[] {
            if(!this.parent) {
                return [];
            }

            return this.parent.getKeys();
        }
        
        getKeys(): string[] {
            const mine = Object.keys(this.table);
            const theirs = this.getParentKeys();

            const both = new Set([...mine, ...theirs]);

            return Array.from(both);
        }
    }

    return new VariableTable();
}

const variableBuilder: IRegisterable = {
    builder: () => buildVariableTable(),
    name: 'variableTable',
    dependencies: [],
    singleton: true
};

export {
    variableBuilder,
};
