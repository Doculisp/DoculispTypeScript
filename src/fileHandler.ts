import { IRegisterable } from "./types.containers";
import { IFileHandler } from "./types.fileHandler";

import { IUtil, Result } from "./types.general";

function buildLoader(util: IUtil, fs: any): IFileHandler {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    return {
        load
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileHandler',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};