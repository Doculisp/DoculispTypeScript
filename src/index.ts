#! /usr/bin/env node
import { container } from './container';
import { Command } from 'commander';
import { Result } from './types/types.general';
import { IController } from './types/types.controller';
import { IVersion } from './types/types.version';
import figlet from 'figlet';

const program = new Command();
const controller = container.buildAs<IController>('controller');
const versionGetter = container.buildAs<IVersion>('version');

const versionMaybe = versionGetter.getVersion();


let version = '?.?.?'

if(!versionMaybe.success) {
    console.log('doculisp cannot determine its version')
} else {
    version = versionMaybe.value;
}

const helptext = '\n\n' + figlet.textSync('doculisp', {
    font: 'Cybermedium'
}) + `\n                     Version: ${version}`;

program.addHelpText('beforeAll', helptext);
program.addHelpText('afterAll', '\n\n');

program
    .name('doculisp')
    .description('A compiler for markdown')
    .version(version)
    .requiredOption('-s, --source <source_path>', 'the source file to compile')
    .requiredOption('-d, --output <output_path>', 'the output document path for the compiled markdown')
    .option('-t, --test', 'runs the compiler without generating the output file.')
    .parse(process.argv);

function reportResult(result: Result<string | false>) {
    if (result.success) {
        const message = !!result.value ? `Ok! ${result.value} successfully written` : 'Ok!';
        console.log(message);
    }
    else {
        console.log(JSON.stringify(result, null, 4));
    }
}

const options = program.opts();

const sourcePath: string = options['source'];

if (options['test']) {
    const result = controller.test(sourcePath);
    reportResult(result);
}
else {
    const outputPath: string = options['output'];
    const result = controller.compile(sourcePath, outputPath);
    reportResult(result);
}
