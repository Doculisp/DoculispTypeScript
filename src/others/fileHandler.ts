import { IRegisterable } from "../types/types.containers";
import { IFileHandler } from "../types/types.fileHandler";
import { IUtil, Result } from "../types/types.general";

function buildLoader(util: IUtil, fs: any): IFileHandler {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    function write(path: string, text: Result<string>): Result<string> {
        if(!text.success) {
            return text;
        }

        const output = text.value;

        try {
            fs.writeFileSync(path, output, {encoding: 'utf8'});
            return text;
        }
        catch(error) {
            return util.fail(`${error}`, path);
        }
    }

    function getProcessWorkingDirectory(): string {
        return process.cwd();
    }

    function setProcessWorkingDirectory(directory: string): void {
        process.chdir(directory);
    }

    return {
        load,
        write,
        getProcessWorkingDirectory,
        setProcessWorkingDirectory,
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileHandler',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};