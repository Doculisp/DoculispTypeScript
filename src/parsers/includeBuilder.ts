import { IDoculisp, IDoculispParser, ISectionWriter } from "../types/types.astDoculisp";
import { IIncludeBuilder } from "../types/types.includeBuilder";
import { IRegisterable } from "../types/types.containers";
import { DocumentParser } from "../types/types.document";
import { IFileHandler } from "../types/types.fileHandler";
import { IProjectLocation, IUtil, Result } from "../types/types.general";
import { TokenFunction } from "../types/types.tokens";

function buildAstBuilder(util: IUtil, astParse: IDoculispParser, documentParse: DocumentParser, tokenizer: TokenFunction, fileHandler: IFileHandler, path: any) : IIncludeBuilder {

    function _parse(filePath: string, location: IProjectLocation): Result<IDoculisp> {
        const workingDir = fileHandler.getProcessWorkingDirectory();
        const targetDir = path.resolve(path.dirname(filePath));
        try {
            fileHandler.setProcessWorkingDirectory(targetDir);

            const fileMaybe = fileHandler.load(path.basename(filePath));
            if(!fileMaybe.success) {
                return fileMaybe;
            }
    
            const documentResult = documentParse(fileMaybe.value, location);
            const tokens = tokenizer(documentResult);
            const doculisp = astParse.parse(tokens);

            return parseExternals(doculisp);
        }
        finally {
            fileHandler.setProcessWorkingDirectory(workingDir);
        }
    }

    function parse(filePath: string): Result<IDoculisp> {
        return _parse(filePath, { documentDepth: 1, documentIndex: 1, documentPath: filePath });
    }

    function parseSection(doculisp: ISectionWriter): Result<ISectionWriter> {
        for (let index = 0; index < doculisp.include.length; index++) {
            const load = doculisp.include[index];
            if(!load) {
                continue;
            }

            if(load.document) {
                parseSection(load.document);
                continue;
            }

            const astResult = _parse(load.path, { documentDepth: doculisp.documentOrder.documentDepth + 1, documentIndex: index + 1, documentPath: load.path});
            if(!astResult.success) {
                return astResult;
            }

            const astDocument = astResult.value;
            if(astDocument.section.type === 'doculisp-empty') {
                continue;
            }

            load.document = astDocument.section;
        }

        return util.ok(doculisp);
    }

    function parseExternals(astResult: Result<IDoculisp>): Result<IDoculisp> {
        if(!astResult.success) {
            return astResult;
        }

        const doculisp = astResult.value;
        if(doculisp.section.type === 'doculisp-empty') {
            return astResult;
        }

        const newSection = parseSection(doculisp.section);

        if(!newSection.success) {
            return newSection;
        }

        doculisp.section = newSection.value;

        return util.ok(doculisp);
    }

    return {
        parse,
        parseExternals,
    };
}

const astBuilder : IRegisterable = {
    builder: (util: IUtil, astParse: IDoculispParser, documentParse: DocumentParser, tokenizer: TokenFunction, fileHandler: IFileHandler, path: any) => buildAstBuilder(util, astParse, documentParse, tokenizer, fileHandler, path),
    name: 'includeBuilder',
    singleton: true,
    dependencies: ['util', 'astDoculispParse', 'documentParse', 'tokenizer', 'fileHandler', 'path']
};

export {
    astBuilder
};