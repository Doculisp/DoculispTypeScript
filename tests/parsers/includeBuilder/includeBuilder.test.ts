import { Options } from "approvals/lib/Core/Options";
import { IContainer, IDictionary, ITestableContainer } from "../../../src/types/types.containers";
import { IFail, IProjectLocation, ISuccess, IUtil, Result } from "../../../src/types/types.general";
import { IDoculisp, IEmptyDoculisp } from "../../../src/types/types.astDoculisp";
import { IIncludeBuilder } from "../../../src/types/types.includeBuilder";
import { getVerifier } from "../../tools";
import { configure } from "approvals/lib/config";
import { containerPromise } from "../../../src/moduleLoader";
import { IDirectoryHandler, IFileLoader } from "../../../src/types/types.fileHandler";
import { buildProjectLocation, buildPath, testable } from '../../testHelpers';
import { IVariableTestable } from '../../../src/types/types.variableTable';
import { IPath } from '../../../src/types/types.filePath';

describe('includeBuilder', () => {
    let verifyAsJson: (data: any, options?: Options) => void;

    let container: IContainer = null as any;
    let util: IUtil = undefined as any;
    let ok: (successfulValue: any) => ISuccess<any> = undefined as any;
    let fail: (message: string, documentPath?: IPath) => IFail = undefined as any;
    let addPathResult: (filePath: string, result: Result<string>) => void = undefined as any;
    let variableSaver: IVariableTestable = undefined as any;

    beforeAll(() => {
        verifyAsJson = getVerifier(configure);
    });

    function setup(environment: ITestableContainer) {
        util = null as any;
        util = environment.buildAs<IUtil>('util');

        variableSaver = environment.buildAs<IVariableTestable>('variableTable');
        variableSaver.clear();

        ok = util.ok;
        fail = util.fail;

        let pathToResult: IDictionary<Result<string>> = undefined as any;
        pathToResult = null as any;
        pathToResult = {};
        addPathResult = (filePath: string, result: Result<string>): void => {
            pathToResult[filePath] = result;
        }
        const fileHandler: IFileLoader & IDirectoryHandler = {
            load(filePath: IPath): Result<string> {
                const result = pathToResult[filePath.fullName];
                if(result) {
                    return result;
                }

                return util.fail(`filePath has not been setup.`, filePath);
            },
            getProcessWorkingDirectory(): Result<IPath> { return util.ok(buildPath('./', false)); },
            setProcessWorkingDirectory(): Result<undefined> { return util.ok(undefined); },
        };

        environment.replaceBuilder(() => fileHandler, [], 'fileHandler', false);
        environment.replaceValue(buildPath, 'pathConstructor');
    }

    describe('includeParse', () => {
        let toExternalResult: (text: string, projectLocation: IProjectLocation) => Result<IDoculisp | IEmptyDoculisp> = undefined as any;

        beforeEach(async () => {
            container = await containerPromise;
            toExternalResult = testable.include.includeResultBuilder(container, setup);
        });

        it('should handle an empty ast', () => {
            const result = toExternalResult("", buildProjectLocation('C:/_main.dlisp', 1, 1));
            verifyAsJson(result);
        });

        it('should return an error if given an error', () => {
            const builder: IIncludeBuilder = testable.include.parserBuilder(container, setup);

            const expectedResult = fail('This is a failure',  buildPath('M:/y/pah.md'));
            expect(builder.parseExternals(expectedResult, variableSaver)).toBe(expectedResult);
        });

        it('should return an error if there is a file error', () => {
            const badPath = buildPath('B:/add.md');
            const expectedResult = fail('baad file error', badPath);
            addPathResult(badPath.fullName, expectedResult);

            const doc = `<!--
(dl
    (section-meta
        (title A journey down a bad filePath)
        (include
            (Section ${badPath})
        )
    )
)
-->

(content)
`;

            const result = toExternalResult(doc, buildProjectLocation('C:/_main.md', 1, 1));
            expect(result).toBe(expectedResult);
        });

        it('should parse one included document', () =>{
            const subDocument = `<!--
(dl
    (section-meta
        (title Fist Sub-Document)
    )
)
-->

Hello world!
`;
            const subDocumentPath = 'C:/sub.md';

            addPathResult(subDocumentPath, ok(subDocument));

            const document = `
(section-meta
    (title Main Document)
    (include
        (Section ${subDocumentPath})
    )
)

(content (toc numbered-labeled))
`;

            verifyAsJson(toExternalResult(document, buildProjectLocation('C:/_main.dlisp', 1, 1)));
        });

        it('should not an included document not a markdown or dlisp', () =>{
            const subDocument = `<h1>Hello World</h1>`;
            const subDocumentPath = 'C:/sub.docx';

            addPathResult(subDocumentPath, ok(subDocument));

            const document = `
(section-meta
    (title Main Document)
    (include
        (Section ${subDocumentPath})
    )
)

(content (toc numbered-labeled))
`;

            verifyAsJson(toExternalResult(document, buildProjectLocation('C:/_main.dlisp', 1, 1)));
        });

        it('should parse two sub documents', () => {
            const subA = `<!--
(dl
    (section-meta
        (title Sub A)
    )
)
-->

Sub document A text.
`;

            const subAPath = 'A:/subA.md';

            const subB = `<!--
(dl
    (section-meta
        (title Sub B)
    )
)
-->

Sub document B text.
`;
            const subBPath = 'B:/subB.md';

            addPathResult(subAPath, ok(subA));
            addPathResult(subBPath, ok(subB));

            const doc = `
(section-meta
    (title Main Multiple Sub Documents)
    (include
        (Section ${subAPath})
        (Section ${subBPath})
    )
)

(content)
`;

            const result = toExternalResult(doc, buildProjectLocation('_main.dlisp', 1, 1));
            verifyAsJson(result);
        });

        it('should parse a sub document containing a sub document', () => {
            const grandChildDocument = `<!--
(dl
    (section-meta
        (title A Grandchild Document)
    )
)
-->

Hi gramps!
`;
            const grandChildPath = 'C:/grand.md';

            const childDoc = `<!--
(dl
    (section-meta
        (title A Child Document)
        (include
            (GrandChild ${grandChildPath})
        )
    )
)
-->

Hi Dad!

(content)
`;
            const childPath = './child.md';

            addPathResult(grandChildPath, ok(grandChildDocument));
            addPathResult(childPath, ok(childDoc));

            const doc = `<!--
(dl
    (section-meta
        (title Top Level Document)
        (include
            (Child ${childPath})
        )
    )
)
-->
Hello World!

(content)
`;

            const result = toExternalResult(doc, buildProjectLocation('_main.md', 1, 1));
            verifyAsJson(result);
        });
    });

    describe('parse', () => {
        let toResult: (filePath: string) => Result<IDoculisp | IEmptyDoculisp> = undefined as any

        beforeEach(() => {
            const builder = testable.include.resultBuilder(container, setup);
            toResult = (filePath) => builder(buildPath(filePath));
        });

        it('should return file error if there is one', () => {
            const docPath = 'C:/bad.md';

            const expectedResult = fail('baad file', buildPath(docPath));

            addPathResult(docPath, expectedResult);

            const result = toResult(docPath);
            expect(result).toBe(expectedResult);
        });

        it('should parse an empty file', () => {
            const doc = '';
            const docPath = 'empty.md';
            
            addPathResult(docPath, ok(doc));

            const result = toResult(docPath);
            verifyAsJson(result);
        });

        it('should parse a document with a child and grand child', () => {
            const grandChildPath = './grandchild.md'
            const grandchild = `
<!--
(dl
    (section-meta
        (title The GRoot of all Docs)
    )
)
-->

Hello from the grand child
`;

            const childPath = './childDoc.md'
            const child = `
<!--
(dl
    (section-meta
        (title The Child to Bare)
        (include
            (Chip ${grandChildPath})
        )
    )
)
-->
hello from the child

<!-- (dl (content)) -->
`;

            addPathResult(grandChildPath, ok(grandchild));
            addPathResult(childPath, ok(child));

            const docPath = './_main.dlisp';
            const doc = `
(section-meta
    (title The True Root)
    (include
        (Block ${childPath})
    )
)

(content (toc))
`;

            addPathResult(docPath, ok(doc));

            const result = toResult(docPath);

            verifyAsJson(result);
        });
    });
});