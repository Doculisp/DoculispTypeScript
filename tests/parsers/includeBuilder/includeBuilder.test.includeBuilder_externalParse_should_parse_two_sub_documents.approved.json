{
  "success": true,
  "value": {
    "projectLocation": {
      "documentDepth": 1,
      "documentIndex": 1,
      "documentPath": "_main.dlisp"
    },
    "section": {
      "doculisp": [
        {
          "documentOrder": {
            "char": 2,
            "documentDepth": 1,
            "documentIndex": 1,
            "documentPath": "_main.dlisp",
            "line": 2
          },
          "label": "# Main Multiple Sub Documents #",
          "ref_link": "#main-multiple-sub-documents",
          "title": "Main Multiple Sub Documents",
          "type": "doculisp-title"
        },
        {
          "documentOrder": {
            "char": 2,
            "documentDepth": 1,
            "documentIndex": 1,
            "documentPath": "_main.dlisp",
            "line": 10
          },
          "type": "doculisp-content"
        }
      ],
      "documentOrder": {
        "char": 1,
        "documentDepth": 1,
        "documentIndex": 1,
        "documentPath": "_main.dlisp",
        "line": 1
      },
      "include": [
        {
          "document": {
            "doculisp": [
              {
                "documentOrder": {
                  "char": 2,
                  "documentDepth": 2,
                  "documentIndex": 1,
                  "documentPath": "A:/subA.md",
                  "line": 2
                },
                "label": "## Sub A ##",
                "ref_link": "#sub-a",
                "title": "Sub A",
                "type": "doculisp-title"
              },
              {
                "documentOrder": {
                  "char": 1,
                  "documentDepth": 2,
                  "documentIndex": 1,
                  "documentPath": "A:/subA.md",
                  "line": 9
                },
                "type": "doculisp-write",
                "value": "Sub document A text."
              }
            ],
            "documentOrder": {
              "char": 1,
              "documentDepth": 2,
              "documentIndex": 1,
              "documentPath": "A:/subA.md",
              "line": 1
            },
            "include": [],
            "type": "doculisp-section"
          },
          "documentOrder": {
            "char": 10,
            "documentDepth": 1,
            "documentIndex": 1,
            "documentPath": "_main.dlisp",
            "line": 5
          },
          "path": "A:/subA.md",
          "sectionLabel": "Section",
          "type": "doculisp-load"
        },
        {
          "document": {
            "doculisp": [
              {
                "documentOrder": {
                  "char": 2,
                  "documentDepth": 2,
                  "documentIndex": 2,
                  "documentPath": "B:/subB.md",
                  "line": 2
                },
                "label": "## Sub B ##",
                "ref_link": "#sub-b",
                "title": "Sub B",
                "type": "doculisp-title"
              },
              {
                "documentOrder": {
                  "char": 1,
                  "documentDepth": 2,
                  "documentIndex": 2,
                  "documentPath": "B:/subB.md",
                  "line": 9
                },
                "type": "doculisp-write",
                "value": "Sub document B text."
              }
            ],
            "documentOrder": {
              "char": 1,
              "documentDepth": 2,
              "documentIndex": 2,
              "documentPath": "B:/subB.md",
              "line": 1
            },
            "include": [],
            "type": "doculisp-section"
          },
          "documentOrder": {
            "char": 10,
            "documentDepth": 1,
            "documentIndex": 1,
            "documentPath": "_main.dlisp",
            "line": 6
          },
          "path": "B:/subB.md",
          "sectionLabel": "Section",
          "type": "doculisp-load"
        }
      ],
      "type": "doculisp-section"
    }
  }
}
